### PR TITLE
add delete state of row block v2 for performance

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. BetaRowset do not set block_status of RowBlock corrently, which will incur more compute cost in reader. This patch will set block_status for BetaRowset.
2. the rowset_meta() of Rowset return shared_ptr reference instead of shared_ptr object for performance
3. related issue #2040

### CODE REVIEW

**Comment by kangpinghuang** [comment]
#2040

**Comment by kangpinghuang** [comment]
this improment will reduce my test query time cost by 200ms(total 1.2s).

my test case:
table: compreh_interest
model: dup
record number:14645005
query:
select count(*) from compreh_interest;

original time: 1.2s
optimized time: 1.0s

File: be/src/olap/rowset/segment_v2/column_reader.h [hunk 1]
```diff
@@ -91,41 +92,35 @@ class ColumnReader {
...
-
     PagePointer get_dict_page_pointer() const { return _meta.dict_page(); }
 
-private:
-    ColumnReader(const ColumnReaderOptions& opts,
-                 const ColumnMetaPB& meta,
-                 uint64_t num_rows,
-                 RandomAccessFile* file);
-    Status init();
+    ColumnZoneMap* column_zone_map() {
```
Comment (gaodayue) [inline be/src/olap/rowset/segment_v2/column_reader.h]: I don't like the idea to make the internal member public in order to implement `get_row_ranges_by_zone_map` in FileColumnIterator, this breaks encapsulation. I think you can achieve the same goal by adding an output parameter `std::vector<uint8_t>* _page_delete_status` to `get_row_ranges_by_zone_map`.

File: be/src/olap/rowset/segment_v2/column_reader.h [hunk 2]
```diff
@@ -232,6 +244,9 @@ class FileColumnIterator : public ColumnIterator {
 
     // current rowid
     rowid_t _current_rowid = 0;
+
+    // page indexes those are DEL_PARTIAL_SATISFIED
+    std::unordered_set<uint32_t> _delete_partial_statisfied_pages;
```
Comment (gaodayue) [inline be/src/olap/rowset/segment_v2/column_reader.h]: `unordered_set` here seems to be overkill, isn't `std::vector<uint8_t> _page_delete_status` simpler to deal with?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 3]
```diff
@@ -52,13 +52,14 @@ SegmentIterator::~SegmentIterator() {
 
 Status SegmentIterator::_init() {
     DorisMetrics::segment_read_total.increment(1);
+    RETURN_IF_ERROR(_init_column_iterators());
```
Comment (gaodayue) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp]: also update `_prepare_seek`?

File: be/src/olap/generic_iterators.cpp [hunk 4]
```diff
@@ -139,6 +140,8 @@ class MergeIteratorContext {
     // will return a valid row
     bool valid() const { return _valid; }
 
+    int is_partial_delete() { return _block.delete_state() == DEL_PARTIAL_SATISFIED; }
```
Comment (gaodayue) [inline be/src/olap/generic_iterators.cpp]:
```suggestion
    int is_partial_delete() const { return _block.delete_state() == DEL_PARTIAL_SATISFIED; }
```

**Review by gaodayue — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 3 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp]: hi dayue, here just init return columns, I think there is no need to call _prepare_seek.

File: be/src/olap/generic_iterators.cpp (hunk 4 above)
Comment (kangpinghuang) [reply be/src/olap/generic_iterators.cpp]: ok

File: be/src/olap/rowset/segment_v2/column_reader.h (hunk 2 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/column_reader.h]: ok

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 5]
```diff
@@ -169,12 +170,11 @@ Status SegmentIterator::_get_row_ranges_from_zone_map(RowRanges* zone_map_row_ra
             continue;
         }
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp]: > so SegmentIterator should be implemented fully based on ColumnIterator
I agree with this. So I move get_row_ranges_by_zone_map from ColumnReader to ColumnIterator. And Default column will just return Status::OK and reserve the row ranges.
I will remove friend declare of SegmentIterator from Segment.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 3 above)
Comment (gaodayue) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp]: OK, you're right. I didn't realize that `_init_column_iterators()` creates iterators for return columns only, maybe we should add a comment on `_init_column_iterators` or `_schema`?

File: be/src/olap/rowset/segment_v2/column_reader.h (hunk 1 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/column_reader.h]: ok

File: be/src/olap/generic_iterators.cpp [hunk 6]
```diff
@@ -254,6 +257,7 @@ Status MergeIterator::init(const StorageReadOptions& opts) {
 
 Status MergeIterator::next_batch(RowBlockV2* block) {
     size_t row_idx = 0;
+    block->set_delete_state(DEL_NOT_SATISFIED);
```
Comment (chaoyli) [inline be/src/olap/generic_iterators.cpp]: I think you can set DEL_NOT_SATISFIED default state.

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 7]
```diff
@@ -214,12 +217,15 @@ Status ColumnReader::_get_filtered_pages(CondColumn* cond_column,
         if (cond_column == nullptr || cond_column->eval({min_value.get(), max_value.get()})) {
             bool should_read = true;
             for (auto& col_cond : delete_conditions) {
-                if (col_cond->del_eval({min_value.get(), max_value.get()}) == DEL_SATISFIED) {
+                int state = col_cond->del_eval({min_value.get(), max_value.get()});
```
Comment (chaoyli) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: not use int to indicate it.

File: be/src/olap/generic_iterators.cpp [hunk 8]
```diff
@@ -262,6 +266,9 @@ Status MergeIterator::next_batch(RowBlockV2* block) {
         // copy current row to block
         copy_row(&dst_row, ctx->current_row(), block->pool());
 
+        if (ctx->is_partial_delete()) {
+            block->set_delete_state(DEL_PARTIAL_SATISFIED);
```
Comment (chaoyli) [inline be/src/olap/generic_iterators.cpp]: I can not understand why you set block DEL_PARTIAL_SATISFIED, I think you will saturate the block. All rows of this block are meaningful.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/column_block.h [hunk 9]
```diff
@@ -55,11 +57,20 @@ class ColumnBlock {
 
     size_t nrows() const { return _nrows; }
 
+    void set_delete_state(uint8_t delete_state) {
+        if (delete_state > _delete_state) {
+            _delete_state = delete_state;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/column_block.h]: ok, I will fix it.

File: be/src/olap/generic_iterators.cpp (hunk 8 above)
Comment (kangpinghuang) [reply be/src/olap/generic_iterators.cpp]: Here the block is not filtered yet, should should be set delete status.
Delete can be calculated here, but need to refactor condition to realize it. which will be done in another pr.

**Review by chaoyli — APPROVED** [review]

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**